### PR TITLE
IMotor

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.util.function.BiFunction;
+
+public class SparkMaxMotor implements IMotor {
+
+	protected final CANSparkMax motor;
+	protected final BiFunction<Rotation2d, Rotation2d, Rotation2d> feedforward;
+	protected final SparkMaxConstants constants;
+
+	public SparkMaxMotor(CANSparkMax motor, BiFunction<Rotation2d, Rotation2d, Rotation2d> feedforward, SparkMaxConstants constants) {
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java:20]: why get feed forward as separate parameter if you have it in the constants?

File: src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	public void setVoltage(double voltage) {
+		motor.setVoltage(voltage);
+	}
+
+	@Override
+	public void setTargetVelocity(CloseLoopControl velocityControl) {
+		motor.getPIDController()
+			.setReference(
+				velocityControl.targetSetPoint().getRotations(),
+				CANSparkBase.ControlType.kVelocity,
```
Comment (drorta) [inline src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java:57]: can be saved as member or constant but not really needed

File: src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	@Override
+	public void updateInputs(MotorInputsAutoLogged motorInputs) {
+		motorInputs.connected = true;
+		motorInputs.angle = Rotation2d.fromRotations(motor.getEncoder().getPosition());
+		motorInputs.velocity = Rotation2d.fromRotations(motor.getEncoder().getVelocity());
+		motorInputs.current = motor.getOutputCurrent();
+		motorInputs.voltage = motor.getBusVoltage() * motor.getAppliedOutput();
+	}
+
+}
```
Comment (drorta) [inline src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java:96]: space after class ends

**Review by bossyshrimp68 — DISMISSED** [review]

File: src/main/java/frc/robot/hardware/talonfx/TalonFXMotor.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import frc.robot.hardware.IMotor;
+import frc.robot.hardware.MotorInputsAutoLogged;
+import frc.utils.calibration.sysid.SysIdCalibrator;
+import frc.utils.devicewrappers.TalonFXWrapper;
+
+public class TalonFXMotor implements IMotor {
+
+	protected final TalonFXWrapper motor;
+	protected final TalonFXSignals signals;
+	protected final SysIdCalibrator.SysIdConfigInfo sysidConfigInfo;
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/talonfx/TalonFXMotor.java:17]: same

File: src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import frc.robot.hardware.MotorInputsAutoLogged;
+import frc.utils.calibration.sysid.SysIdCalibrator;
+
+import java.util.function.BiFunction;
+
+public class SparkMaxMotor implements IMotor {
+
+	protected final CANSparkMax motor;
+	protected final SparkMaxConstants constants;
+	protected final BiFunction<Rotation2d, Rotation2d, Rotation2d> feedforward;
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/hardware/cansparkmax/SparkMaxMotor.java:18]: ירושה

File: src/main/java/frc/robot/hardware/talonfx/TalonFXMotor.java (hunk 4 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/talonfx/TalonFXMotor.java:17]: ירושה

File: src/main/java/frc/robot/hardware/CloseLoopControl.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package frc.robot.hardware;
+
+import com.ctre.phoenix6.controls.ControlRequest;
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public record CloseLoopControl(ControlRequest controlRequest, int pidSlot, Rotation2d targetSetPoint, ControlState controlState) {}
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/hardware/CloseLoopControl.java:6]: how? like what do you mean

[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/hardware/CloseLoopControl.java:6]: but they do, you must have an object that knows to return slot, setpoint and control

[the remark this answers is not in the document]
Comment (hillelv) [reply src/main/java/frc/robot/hardware/CloseLoopControl.java:6]: in talonFX you only use the control request in the set target

**Review by YoniKiriaty — DISMISSED** [review]

File: src/main/java/frc/robot/hardware/motor/PIDAble.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package frc.robot.hardware.motor;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.request.angle.IAngleRequest;
+import frc.robot.hardware.request.value.IValueRequest;
+import frc.utils.calibration.sysid.SysIdCalibrator;
+
+public interface PIDAble {
```
Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/motor/PIDAble.java:8]: be consistent. its either
IMotor and IPID 
or
MotorAble (or smth) and PIDAble

File: src/main/java/frc/robot/hardware/motor/sparkmax/SparkMaxMotor.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+import frc.robot.hardware.motor.PIDAble;
+import frc.robot.hardware.motor.PIDAbleInputsAutoLogged;
+import frc.robot.hardware.motor.ProfileAble;
+import frc.robot.hardware.request.angle.IAngleRequest;
+import frc.robot.hardware.request.angle.SparkMaxAngleRequest;
+import frc.robot.hardware.request.value.IValueRequest;
+import frc.robot.hardware.request.value.SparkMaxValueRequest;
+import frc.utils.calibration.sysid.SysIdCalibrator;
+
+public class SparkMaxMotor implements IMotor, PIDAble, ProfileAble {
```
Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/motor/sparkmax/SparkMaxMotor.java:17]: rename to BrushlessCANSparkMax

File: src/main/java/frc/robot/hardware/request/angle/IAngleRequest.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package frc.robot.hardware.request.angle;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public interface IAngleRequest {
```
Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/request/angle/IAngleRequest.java:5]: maby make another interface for position (linear position in meters) or is it the IValueRequest

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/motor/PIDAble.java (hunk 7 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/motor/PIDAble.java:8]: its consistent, the unterface represent what it does.
imotor implements a motor, so ipid should implement pid. Butt its not the case so Pidable represent something that can do pid

File: src/main/java/frc/robot/hardware/request/angle/IAngleRequest.java (hunk 9 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/request/angle/IAngleRequest.java:5]: its the ivalue, now it was made for volt and current

File: src/main/java/frc/robot/hardware/motor/sparkmax/SparkMaxMotor.java (hunk 8 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/motor/sparkmax/SparkMaxMotor.java:17]: bruhhhh

**Pull request closed without merging**